File: lab9oops.py
import math

class Vehicle:
    def __init__(self, make, model):
        self.make = make
        self.model = model
    
    def display(self):
        print(f"Vehicle: {self.make} {self.model}")
    
class Car(Vehicle):
    def __init__(self, make, model, fuel_type):
        super().__init__(make, model)
        self.fuel_type = fuel_type

    def display(self):
        print(f"Car: {self.make} {self.model}, Fuel Type: {self.fuel_type}")

    def fuel_efficiency(self, mileage=None):
        if mileage is None:
            print("Fuel efficiency data not provided.")
        else:
            print(f"Car mileage: {mileage} km/l")
        
class Truck(Vehicle):
    def __init__(self, make, model, load_capacity):
        super().__init__(make, model)
        self.load_capacity = load_capacity
    
    def display(self):
        print(f"Truck: {self.make} {self.model}, Load Capacity: {self.load_capacity} tons")

    def fuel_efficiency(self, load=None):
        if load is None:
            print("Fuel efficiency data not provided.")
        else:
            print(f"Truck fuel efficiency for load {load} tons: 8 km/l")

vehicle = Vehicle("Generic", "Model X")
car = Car("Toyota", "Corolla", "Petrol")
truck = Truck("Volvo", "FH", 10)

vehicle.display()
car.display()
truck.display()

car.fuel_efficiency()
car.fuel_efficiency(15)

truck.fuel_efficiency()
truck.fuel_efficiency(5)

File: lab9oops.java
class Vehicle {
    String make;
    String model;

    public Vehicle(String make, String model) {
        this.make = make;
        this.model = model;
    }

    public void display() {
        System.out.println("Vehicle: " + make + " " + model);
    }
}

class Car extends Vehicle {
    String fuelType;

    public Car(String make, String model, String fuelType) {
        super(make, model);
        this.fuelType = fuelType;
    }

    @Override
    public void display() {
        System.out.println("Car: " + make + " " + model + ", Fuel Type: " + fuelType);
    }

    public void fuelEfficiency() {
        System.out.println("Fuel efficiency data not provided.");
    }

    public void fuelEfficiency(int mileage) {
        System.out.println("Car mileage: " + mileage + " km/l");
    }
}

class Truck extends Vehicle {
    int loadCapacity;

    public Truck(String make, String model, int loadCapacity) {
        super(make, model);
        this.loadCapacity = loadCapacity;
    }

    @Override
    public void display() {
        System.out.println("Truck: " + make + " " + model + ", Load Capacity: " + loadCapacity + " tons");
    }

    public void fuelEfficiency() {
        System.out.println("Fuel efficiency data not provided.");
    }

    public void fuelEfficiency(int load) {
        System.out.println("Truck fuel efficiency for load " + load + " tons: 8 km/l");
    }
}

public class lab9oops {
    public static void main(String[] args) {
        Vehicle vehicle = new Vehicle("Generic", "Model X");
        Car car = new Car("Toyota", "Corolla", "Petrol");
        Truck truck = new Truck("Volvo", "FH", 10);

        vehicle.display();
        car.display();
        truck.display();

        car.fuelEfficiency();
        car.fuelEfficiency(15);

        truck.fuelEfficiency();
        truck.fuelEfficiency(5);
    }
}
